src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Html/DefaultHtmlCodeActionResolver.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal class DefaultHtmlCodeActionResolver : HtmlCodeActionResolver
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public DefaultHtmlCodeActionResolver(
+            DocumentContextFactory documentContextFactory,
+            ClientNotifierServiceBase languageServer,
+            RazorDocumentMappingService documentMappingService)
+            : base(languageServer)
+        {
+            if (documentContextFactory is null)
+            {
+                throw new ArgumentNullException(nameof(documentContextFactory));
+            }
+
+            if (documentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingService));
+            }
+
+            _documentContextFactory = documentContextFactory;
+            _documentMappingService = documentMappingService;
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.Default;
+
+        public async override Task<CodeAction> ResolveAsync(
+            CodeActionResolveParams resolveParams,
+            CodeAction codeAction,
+            CancellationToken cancellationToken)
+        {
+            if (resolveParams is null)
+            {
+                throw new ArgumentNullException(nameof(resolveParams));
+            }
+
+            if (codeAction is null)
+            {
+                throw new ArgumentNullException(nameof(codeAction));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(resolveParams.RazorFileUri, cancellationToken).ConfigureAwait(false);
```

Comment:
Come to think of it, is this the proper abstraction? Shouldn't the caller already have the `documentContext`? Since the CodeAction methods are non-mutating the document may have changed under the covers making this dangerous, and anyway if we already have the documentContext higher up re-computing it is just a bit of a waste.

I would say that anywhere outside of initial request-handling that we import `DocumentContextFactory` should get a second look.

Reply:
You make a good point, but I think I'll do this in a follow up, rather than make this PR any bigger. The problem already exists in the C# code action code too, so will clean it all up together. Logged https://github.com/dotnet/razor/issues/7926

Reply:
Update: The code action resolve handler is documentless!